lib/Parse/Lexer.cpp

```diff
@@ -1373,6 +1373,64 @@ void Lexer::lexEscapedIdentifier() {
   formToken(tok::backtick, Quote);
 }
 
+/// Find the end of a version control conflict marker.
+static const char *findConflictEnd(const char *CurPtr, const char *BufferEnd,
+                                   ConflictMarkerKind CMK) {
+  const char *terminator = CMK == ConflictMarkerKind::Perforce ? "<<<<\n"
+                                                               : ">>>>>>>";
+  size_t termLen = CMK == ConflictMarkerKind::Perforce ? 5 : 7;
+  
+  // Get a reference to the rest of the buffer minus the length of the start
+  // of the conflict marker.
+  auto restOfBuffer = StringRef(CurPtr, BufferEnd - CurPtr).substr(termLen);
+  size_t endPos = restOfBuffer.find(terminator);
+  while (endPos != StringRef::npos) {
+    // Must occur at start of line.
+    if (endPos == 0 ||
+        (restOfBuffer[endPos - 1] != '\r' && restOfBuffer[endPos - 1] != '\n'))
+    {
+      restOfBuffer = restOfBuffer.substr(endPos+termLen);
+      endPos = restOfBuffer.find(terminator);
+      continue;
+    }
+    return restOfBuffer.data() + endPos;
+  }
+  return nullptr;
+}
+
+/// Try to lex conflict markers by checking for the presence of the start and
+/// end of the marker in diff3 or Perforce style respectively.
+bool Lexer::tryLexConflictMarker() {
+  const char *Ptr = CurPtr - 1;
+
+  // Only a conflict marker if it starts at the beginning of a line.
+  if (Ptr != BufferStart && Ptr[-1] != '\n' && Ptr[-1] != '\r')
+    return false;
+  
+  // Check to see if we have <<<<<<< or >>>>.
+  if (((BufferEnd - Ptr) < 8 || StringRef(Ptr, 7) != "<<<<<<<") &&
+      ((BufferEnd - Ptr) < 6 || StringRef(Ptr, 5) != ">>>> "))
```

Comment:
Hm. But that's not a concern for `<<<<<<<`? (I guess it is a little less likely, and Perforce markers are probably guaranteed to have the space.)